Conversation custom sentences (#25844)
home-assistant/home-assistant.github.io: source/_integrations/conversation.markdown

conversation:
```

## Default sentences

By default, a collection of [community contributed sentences](https://github.com/home-assistant/intents/) are supported in a growing [list of languages](https://developers.home-assistant.io/docs/voice/intent-recognition/supported-languages).

> In English, you can say things like "turn on kitchen lights" or "turn off lights in the bedroom" if you have an area named "bedroom". These sentences are automatically mapped to the [built-in intents](https://developers.home-assistant.io/docs/intent_builtin/).

Reviewer: Don't think we have to use `built-in intents`, might confuse.
```suggestion
In English, you can say things like "turn on kitchen lights" or "turn off lights in the bedroom" if you have an area named "bedroom".

## Adding custom sentences

> You can add your own sentence templates to teach Home Assistant about new sentences. These sentences can work with the [built-in intents](https://developers.home-assistant.io/docs/intent_builtin/) or trigger a custom action by defining custom intents with the [intent script integration](/integrations/intent_script/).

Reviewer: ```suggestion
You can add your own [sentence templates](https://developers.home-assistant.io/docs/voice/intent-recognition/template-sentence-syntax) to teach Home Assistant about new sentences. These sentences can work with the [built-in intents](https://developers.home-assistant.io/docs/intent_builtin/) or trigger a custom action by defining custom intents with the [intent script integration](/integrations/intent_script/).
```

Reviewer: We also link it in the advanced section. Not sure if we need to link it here.

> Custom sentences can be mapped to your own [intents](https://developers.home-assistant.io/docs/intent_index) and then linked to any action in Home Assistant with the [intent script integration](/integrations/intent_script/).

Reviewer: ```suggestion
You can add your own sentence templates to teach Home Assistant about new sentences. These sentences can work with the [built-in intents](https://developers.home-assistant.io/docs/intent_builtin/) or trigger a custom action by defining custom intents with the [intent script integration](/integrations/intent_script/).
```

> To get started, create a `custom_sentences/<language>` directory in your Home Assistant `config` directory where `<language>` is the [language code](https://developers.home-assistant.io/docs/voice/intent-recognition/supported-languages) you've configured, such as `en` for English.

Reviewer: ```suggestion
To get started, create a `custom_sentences/<language>` directory in your Home Assistant `config` directory where `<language>` is the [language code](https://developers.home-assistant.io/docs/voice/intent-recognition/supported-languages) of your language, such as `en` for English.
```

For an English example, create the file `config/custom_sentences/en/temperature.yaml` and add:

{% raw %}

```yaml
# Example temperature.yaml entry
language: "en"
intents:

>   OutsideHumidity:

Reviewer: To make it clear it's a custom one

```suggestion
  CustomOutsideHumidity:
```

    data:
      - sentences:
          - "What is the humidity outside"
```

{% endraw %}

> To produce a response, create an `intent_script` entry in your `configuration.yaml` file:

Reviewer: ```suggestion
To teach Home Assistant how to handle the custom `CustomOutsideHumidity ` intent, create an `intent_script` entry in your `configuration.yaml` file:
```

{% raw %}

```yaml
# Example configuration.yaml entry
intent_script:

>   OutsideHumidity:

Reviewer: ```suggestion
  CustomOutsideHumidity:
```

In `config/custom_sentences/en/media_player.yaml`:

{% raw %}

```yaml
# Example media_player.yaml entry
language: "en"
intents:
  SetVolume:
    data:
      - sentences:
          - "(set|change) {media_player} volume to {volume} [percent]"
          - "(set|change) volume in [the] {media_player} to {volume} [percent]"
lists:
  media_player:
    values:
      - in: "living room"
        out: "media_player.living_room"
      - in: "bedroom"
        out: "media_player.bedroom"
  volume:
    range:
      from: 0
      to: 100
```

{% endraw %}

The `media_player` list is referenced as `{media_player}` in the sentence templates. Its values contain both an `in` and `out` part, where `in` is the input text and `out` is the output value (`entity_id`) used in `intent_script` to call the `media_player.volume_set` service.

In `configuration.yaml`:

{% raw %}

```yaml
# Example configuration.yaml entry
intent_script:
  SetVolume:
    action:
      service: "media_player.volume_set"
      data:
        entity_id: "{{ media_player }}"
        volume_level: "{{ volume / 100.0 }}"
    speech:

>       text: "Volume changed to {{ volume }}"

Reviewer: To make the text not contain a fraction

```suggestion
      text: "Volume changed to {{ volume | int }}"
```

Author: `volume` is already an `int` coming from the intent. Is this just in case it ends up being a `float`?

Reviewer: oh right, I was confused with `volume_level`

To get started, create a `custom_sentences/<language>` directory in your Home Assistant `config` directory where `<language>` is the [language code](https://developers.home-assistant.io/docs/voice/intent-recognition/supported-languages) of your language, such as `en` for English. These YAML files are automatically merged, and may contain intents, lists, or expansion rules.

For an English example, create the file `config/custom_sentences/en/temperature.yaml` and add:

{% raw %}

```yaml
# Example temperature.yaml entry
language: "en"
intents:
  CustomOutsideHumidity:
    data:
      - sentences:
          - "What is the humidity outside"
```

{% endraw %}

To teach Home Assistant how to handle the custom `CustomOutsideHumidity` intent, create an `intent_script` entry in your `configuration.yaml` file:

{% raw %}

```yaml
# Example configuration.yaml entry
intent_script:
  CustomOutsideHumidity:
    speech:
      text: "It is currently {{ states("sensor.outside_humidity") }} percent humidity outside."
```

{% endraw %}

> More complex [actions](https://www.home-assistant.io/docs/scripts/) can be done in `intent_script`, such as calling services and firing events.

Reviewer: ```suggestion
More complex [actions](/docs/scripts/) can be done in `intent_script`, such as calling services and firing events.
```

## Extending built-in intents

Extending the built-in intents, such as `HassTurnOn` and `HassTurnOff`, can be done as well.

For example, create the file `config/custom_sentences/en/on_off.yaml` and add:

{% raw %}

```yaml
# Example on_off.yaml entry
language: "en"
intents:
  HassTurnOn:
    data:
      - sentences:

>           - "engage [the] kitchen lights"
>         slots:
>           name: "kitchen lights"

Reviewer: Let's make it more weird.
```suggestion
          - "fire up the rockets"
        slots:
          name: "rocket lamp"
```

>   HassTurnOff:
>     data:
>       - sentences:
>           - "disengage [the] kitchen lights"
>         slots:
>           name: "kitchen lights"

Reviewer: No need to show 2 intents I think? Better keep examples small
```suggestion
```

| Service data attribute | Optional | Description                                      |
|------------------------|----------|--------------------------------------------------|
| `text`                 |      yes | Transcribed text                                 |

## Service `conversation.reload`

| Service data attribute | Optional | Description                                      |
|------------------------|----------|--------------------------------------------------|

> | `language`             |      yes | Language to clear intent cache for               |

Reviewer: ```suggestion
| `language`             |      yes | Language to clear intent cache for. Defaults to Home Assistant language.               |
```
